src/gpio.rs

```diff
@@ -96,6 +153,66 @@ pub trait ExtiPin {
     fn check_interrupt(&mut self) -> bool;
 }
 
+pub enum Dynamic {
+    InputFloating,
+    InputPullUp,
+    InputPullDown,
+    OutputPushPull,
+    OutputOpenDrain,
+}
+
+impl Active for Dynamic {}
+
+#[derive(Debug, PartialEq)]
+pub enum PinModeError {
+    IncorrectMode,
+}
+
+impl Dynamic {
+    fn is_input(&self) -> bool {
+        use Dynamic::*;
+        match self {
+            InputFloating | InputPullUp | InputPullDown | OutputOpenDrain => true,
+            OutputPushPull => false,
+        }
+    }
+    fn is_output(&self) -> bool {
+        use Dynamic::*;
+        match self {
+            InputFloating | InputPullUp | InputPullDown => false,
+            OutputPushPull | OutputOpenDrain => true,
+        }
+    }
+}
+
+/// NOTE: This trait should ideally be private but must be pub in order to avoid
+/// complaints from the compiler.
+pub trait PinMode<CR> {
+    unsafe fn set_mode(cr: &mut CR) -> Self;
+}
+
+// These impls are needed because a macro can not brace initialise a ty token
+impl<MODE> Input<MODE> {
+    fn _new() -> Self {
```

Comment:
`const fn`

Reply:
I made all of these `const`, but I'm not sure how often we should use `inline`. My general feeling is that compilers are usually better at deciding when to inline things than we are, but maybe that's wrong. Would you like any more consts added?

Reply:
#[inline] is just an indication that this might be useful to inline, it just shifts the bias a bit and will be ignored in dev mode altogether.